synapse/lib/db.py

```diff
@@ -0,0 +1,156 @@
+    '''
+    A unified helper for managing db transaction behavior.
+    '''
+    def __init__(self, pool, db):
+        s_eventbus.EventBus.__init__(self)
+
+        self.db = db
+        self.refs = 0
+        self.pool = pool
+        self.curs = None
+        self.lockd = False
+
+    def wrlock(self):
+        while not self.lockd:
+            if self.isfini:
+                raise IsFini()
+
+            self.lockd = self.pool.wlock.acquire(timeout=1)
+
+    def execute(self, *args):
+        '''
+        Execute a query on the cursor for this xact.
+
+        NOTE: This *will* acquire the write-lock for
+              the db Pool() until your transaction is
+              complete.  Use select() for read queries.
+
+        Args:
+            *args (list): The args to hand to cursor execute().
+
+        Returns:
+            Cursor: see DB API
+        '''
+        self.wrlock()
+        return self.curs.execute(*args)
+
+    def update(self, *args):
+        self.wrlock()
+        return self.execute(*args).rowcount
+
+    def executemany(self, *args):
+        self.wrlock()
+        return self.curs.executemany(*args)
+
+    def select(self, *args):
+        '''
+        Read-only optimized cursor execute method.
+
+        Args:
+            *args (list): The args to hand to cursor execute().
+
+        Returns:
+            Cursor: see DB API
+        '''
+        if self.isfini:
+            raise IsFini()
+
+        return self.curs.execute(*args)
+
+    def commit(self):
+        '''
+        Perform an incremental commit().
+
+        See DB API
+        '''
+        # on an incremental commit, release the wlock
+        self.db.commit()
+        if self.lockd:
+            self.pool.wlock.release()
+            # maybe sleep(0.0000001)?
+            self.pool.wlock.acquire()
+
+    def __enter__(self):
+        self.refs += 1
+
+        if self.refs == 1:
+            self.curs = self.db.cursor()
+
+        return self
+
+    def __exit__(self, exc, cls, tb):
+
+        self.refs -= 1
+        if self.refs == 0:
+
+            self.curs.close()
+            self.db.commit()
+
+            if self.lockd:
+                self.lockd = False
+                self.pool.wlock.release()
+
+            self.curs = None
+            self.pool._put(self)
+
+class Pool(s_eventbus.EventBus):
+    '''
+    The Pool allows generic db connection pooling using
+    a factory/ctor method and a python queue.
+    '''
+    _xact_ctor = Xact
+
+    def __init__(self, size, ctor):
+        s_eventbus.EventBus.__init__(self)
+
+        self.size = size
+        self.iden = s_common.guid()
+
+        self.dbque = s_queue.Queue()
+        self.onfini(self.dbque.fini)
+
+        self.wlock = threading.Lock()
+
+        self.conns = [ctor() for i in range(size)]
+
+        [self._initXact(db) for db in self.conns]
+
+    def _initXact(self, db):
+        xact = self._xact_ctor(self, db)
+        self.dbque.put(xact)
+
+    def _onPoolFini(self):
```

Comment:
@invisig0th - I went to write a test for this and noticed that it is unused. Should it be registered in this `Pool`'s ctor? Should we remove it?